RecoEgamma/EgammaTools/plugins/HGCalElectronIDValueMapProducer.cc

```diff
@@ -0,0 +1,238 @@
+
+  private:
+    void beginStream(edm::StreamID) override;
+    void produce(edm::Event&, const edm::EventSetup&) override;
+    void endStream() override;
+
+    // ----------member data ---------------------------
+    edm::EDGetTokenT<edm::View<reco::GsfElectron>> ElectronsToken_;
+    float radius_;
+    std::map<const std::string, std::vector<float>> maps_;
+
+    std::unique_ptr<ElectronIDHelper> eIDHelper_;
+};
+
+HGCalElectronIDValueMapProducer::HGCalElectronIDValueMapProducer(const edm::ParameterSet& iConfig) :
+  ElectronsToken_(consumes<edm::View<reco::GsfElectron>>(iConfig.getParameter<edm::InputTag>("electrons"))),
+  radius_(iConfig.getParameter<double>("pcaRadius"))
+{
+  // All the ValueMap names to output are defined in the python config
+  // so that potential consumers can configure themselves in a simple manner
+  for(auto key : iConfig.getParameter<std::vector<std::string>>("variables")) {
+    maps_[key] = {};
+    produces<edm::ValueMap<float>>(key);
+  }
+
+  eIDHelper_.reset(new ElectronIDHelper(iConfig, consumesCollector()));
+}
+
+
+HGCalElectronIDValueMapProducer::~HGCalElectronIDValueMapProducer()
+{
+}
+
+
+// ------------ method called to produce the data  ------------
+void
+HGCalElectronIDValueMapProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+  using namespace edm;
+
+  Handle<edm::View<reco::GsfElectron>> ElectronsH;
+  iEvent.getByToken(ElectronsToken_, ElectronsH);
+
+  const size_t prevMapSize = maps_.size();
+
+  // Clear previous map
+  for(auto&& kv : maps_) kv.second.clear();
+
+  // Set up helper tool
+  eIDHelper_->eventInit(iEvent,iSetup);
+
+  for(size_t iEle=0; iEle<ElectronsH->size(); ++iEle) {
+    const auto& electron = ElectronsH->at(iEle);
+
+    if(electron.isEB()) {
+      // Fill some dummy value
+      for(auto&& kv : maps_) {
+	kv.second.push_back(0.);
+      }
+    }
+    else {
+      eIDHelper_->computeHGCAL(electron, radius_);
+
+      // check the PCA has worked out
+      if (eIDHelper_->sigmaUU() == -1){
+	  for(auto&& kv : maps_) {
+	      kv.second.push_back(0.);
+	  }
+	  continue;
+      }
+
+      LongDeps ld(eIDHelper_->energyPerLayer(radius_, true));
+      float measuredDepth, expectedDepth, expectedSigma;
+      float depthCompatibility = eIDHelper_->clusterDepthCompatibility(ld, measuredDepth, expectedDepth, expectedSigma);
+
+      // Fill here all the ValueMaps from their appropriate functions
+
+      // Energies / PT
+      maps_["gsfTrackPt"].push_back(electron.gsfTrack()->pt());
+      maps_["pOutPt"].push_back(std::sqrt(electron.trackMomentumAtVtx().perp2()));
+      maps_["scEt"].push_back(electron.superCluster()->energy() / std::cosh(electron.superCluster()->eta()));
+      maps_["scEnergy"].push_back(electron.superCluster()->energy());
+      maps_["ecOrigEt"].push_back(electron.electronCluster()->energy() / std::cosh(electron.electronCluster()->eta()));
+      maps_["ecOrigEnergy"].push_back(electron.electronCluster()->energy());
```

Comment:
please reduce copy-paste of function calls  : precompute repeated values in temporaries

Reply:
Surely the compiler optimizes this away?

Reply:
It can, but only in some cases - best to do it manually to make sure.